mozilla-mobile/firefox-ios
# Bugfix FXIOS-10216 Fix disappearing custom search engine favicons [Favicon Refactor]

## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10216)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/22360)

## :bulb: Description
Please watch the video attached to the GitHub issue above for a demonstration of the bug being fixed.

This bug only happened when adding new custom search engines for which we have an image bundled in BrowserKit's `BundledTopSitesFavicons.xcassets` asset catalogue (e.g. YouTube, Amazon, Google...).

This deceptively simple bug took a while to track down. The issue is that, after we added additional support to check the bundle prior to fetching favicon images from the internet, `OpenSearchEngine`s were now trying to encode a `UIImage` **backed by the asset catalogue**. This functionality was added during the favicon widget update, and it helpfully ensures we get nice high quality images for YouTube, Google, etc. when those icons would otherwise be low-quality `.ico` files pulled from the internet.

Unfortunately, `UIImage`s backed by the asset catalogue contain the asset name but not the actual underlying `Data`. So prior to encoding `OpenSearchEngine` `UIImage`s for storage in the `customEnglines.plist` file, as a workaround we can force create a new `UIImage` initialized with the raw PNG data.

### Random Note About the Unit Tests
This PR also adds some unit tests and test assets to the ClientTests target. However, the behaviour seen in the tests isn't 100% identical to the app since the bundles from which the images originate are different. It appears the unit test can successfully encode and decode `UIImage`s from just the bundle asset name because the bundle is within the same target. That is why we check that the file's data size is at least greater than 6,000 kB.

## Test Instructions:

- Launch the app
- Go to Settings > Search > Add Search Engine
- Add a new engine for YouTube with the following query string: `https://www.youtube.com/results?search_query=%s`
- Tap `Save`
- Return to Home and start typing in the toolbar to start a search
- Ensure that the YouTube icon appears in the tab strip above the keyboard
- Hard kill the app
- Restart the app
- Ensure that the YouTube icon _still_ appears in the tab strip above the keyboard and is not blank.

<img width="300" alt="Screenshot 2024-10-09 at 3 11 54 PM" src="https://github.com/user-attachments/assets/6c9754db-e747-4594-bc3b-5900e9168aee">

## :pencil: Checklist
You have to check all boxes before merging
